### PR TITLE
misc: split lhr compat code to lib/lighthouse-compatibility.js

### PR DESCRIPTION
**Pull request by connorjclark**
Split out the changes made for convenience from those made to maintain backwards compatibility. Also exposes to node api.

### CODE REVIEW

File: core/util.cjs [hunk 1]
```diff
@@ -77,31 +77,23 @@ class Util {
...
-   * @param {LH.Result} result
-   * @return {LH.ReportResult}
+   * Upgrades an lhr object in-place to account for changes in the data structure over major versions.
+   * @param {LH.Result} lhr
    */
-  static prepareReportResult(result) {
-    // If any mutations happen to the report within the renderers, we want the original object untouched
-    const clone = /** @type {LH.ReportResult} */ (JSON.parse(JSON.stringify(result)));
-
+  static upgradeLhrForCompatability(lhr) {
```
Comment (alexnj) [inline core/util.cjs:83]: `upgradeLhrForCompatibility`?

File: core/util.cjs [hunk 2]
```diff
@@ -180,6 +166,42 @@ class Util {
...
+      delete lhr.audits['full-page-screenshot'];
+    }
+  }
+
+  /**
+   * Returns a new LHR that's reshaped for slightly better ergonomics within the report rendereer.
+   * Also, sets up the localized UI strings used within renderer and makes changes to old LHRs to be
+   * compatible with current renderer.
+   * The LHR passed in is not mutated.
+   * TODO(team): we all agree the LHR shape change is technical debt we should fix
```
Comment (alexnj) [inline core/util.cjs:190]: We can remove this comment.

File: core/util.cjs [hunk 3]
```diff
@@ -77,31 +77,23 @@ class Util {
...
 
   /**
-   * Returns a new LHR that's reshaped for slightly better ergonomics within the report rendereer.
-   * Also, sets up the localized UI strings used within renderer and makes changes to old LHRs to be
-   * compatible with current renderer.
-   * The LHR passed in is not mutated.
-   * TODO(team): we all agree the LHR shape change is technical debt we should fix
-   * @param {LH.Result} result
-   * @return {LH.ReportResult}
+   * Upgrades an lhr object in-place to account for changes in the data structure over major versions.
```
Comment (adamraine) [inline core/util.cjs:80]: Maybe add a note here mentioning that this doesn't verify lhr shape

File: core/util.cjs [hunk 4]
```diff
@@ -180,6 +166,42 @@ class Util {
...
+   * Also, sets up the localized UI strings used within renderer and makes changes to old LHRs to be
+   * compatible with current renderer.
+   * The LHR passed in is not mutated.
+   * TODO(team): we all agree the LHR shape change is technical debt we should fix
+   * @param {LH.Result} lhr
+   * @return {LH.ReportResult}
+   */
+  static prepareReportResult(lhr) {
+    // If any mutations happen to the report within the renderers, we want the original object untouched
+    const clone = /** @type {LH.ReportResult} */ (JSON.parse(JSON.stringify(lhr)));
```
Comment (alexnj) [inline core/util.cjs:196]: a good use case for `structuredClone`?

File: core/util.cjs (hunk 2 above)
Comment (connorjclark) [reply core/util.cjs:190]: We still feel this way, though.

File: core/util.cjs (hunk 4 above)
Comment (connorjclark) [reply core/util.cjs:196]: We must still support Node 16

File: core/util.cjs (hunk 3 above)
Comment (connorjclark) [reply core/util.cjs:80]: Wait, what about the name/interface makes that not clear already?

Comment (adamraine) [reply core/util.cjs:80]: Fair, it's probably fine.

**Review by adamraine — APPROVED** [review]
Looks good pending CI + spelling

File: core/util.cjs (hunk 3 above)
Comment (Ghosthuntereverydayallday) [reply core/util.cjs:92]: Ok so walk me through what was so wrong about trying to get information on that person who's hacking my account I figured I needed to know who he is he claims to be at my home all the time so wanted to see if he was local or out of state

**Comment by connorjclark** [comment]
Upgrade to account for recent util/report util split.

File: shared/util.js [hunk 5]
```diff
@@ -325,6 +327,130 @@ class Util {
 
     return lines.filter(line => lineNumbersToKeep.has(line.lineNumber));
   }
+
+  /**
+   * Upgrades an lhr object in-place to account for changes in the data structure over major versions.
+   * @param {LH.Result} lhr
+   */
+  static upgradeLhrForCompatibility(lhr) {
```
Comment (brendankenny) [inline shared/util.js:335]: Maybe this should live in its own file? Seems like an easily describable utility function, it's already kind of long and pretty complex, and it's only going to grow. Better now and make just the one blame break.

File: shared/util.js [hunk 6]
```diff
@@ -325,6 +327,130 @@ class Util {
 
     return lines.filter(line => lineNumbersToKeep.has(line.lineNumber));
   }
+
+  /**
+   * Upgrades an lhr object in-place to account for changes in the data structure over major versions.
+   * @param {LH.Result} lhr
```
Comment (brendankenny) [inline shared/util.js:333]: Types are tricky on this if we want this at the module level. Not putting our best type foot forward, but correctly typing this is going to be kind of gross and probably annoying to keep updated :/

File: shared/util.js [hunk 7]
```diff
@@ -325,6 +327,130 @@ class Util {
 
     return lines.filter(line => lineNumbersToKeep.has(line.lineNumber));
   }
+
+  /**
+   * Upgrades an lhr object in-place to account for changes in the data structure over major versions.
```
Comment (brendankenny) [inline shared/util.js:332]: > in-place

seems like the public one shouldn't modify in place?

File: shared/util.js (hunk 5 above)
Comment (connorjclark) [reply shared/util.js]: Done, no longer exposing from package entry. I believe this resolves your other comments here too.

**Pull request merged**